01-algorithms/01-algorithms.md

```diff
@@ -50,65 +61,454 @@ The study of computer science is the study of algorithms.  Researchers in CS foc
+    OUTPUT: Boolean indicating whether there is are duplicate elements in the list within k distance
+    '''
+    # get length of input list
+    lst_length = len(numbers)
+
+    # if the input list has less than two elements or k is zero
+    if lst_length < 2 or k == 0:
+      # return false, there cannot be duplicates
+      return False
+    
+    # loop through indices of list
+    for i in range(lst_length):
+      # initialize a second pointer to track end of subarray
+      j = i + 1
+      # initialize a variable to k
+      # will help make sure our subarray stays within size k
+      dist_remaining = k
+      
+      # while we are within k distance from i
+      # and j is not out of bounds of list
+      while dist_remaining > 0 and j < lst_length:
+        # if elements at index i and j are the same
+        if numbers[i] == numbers[j]:
+          # a duplicate within k distance exists
+          return True
+        # otherwise increment j to increase subarray size
+        j += 1
+        # decrement dist_remaining so subarray will stay within
+        # k distance from i
+        dist_remaining -= 1
+    # if we loop through all subarrays and have not found a duplicate
+    # return False
+    return False
+
+  ```
+</details>
+
+This solution works and is a relatively direct straight forward approach. This is often called a **brute force** solution.
+
+### Simplify/Refactor
+
+The first big challenge in software development is to _produce working code_. If we still have time after producing working code, we should consider opportunites to simplify or refactor (improve) our solution.
+
+In the sample solution above we are using a nested loop to repeatedly traverse the list. The outer loop tracked the starting index of our subarrays and the inner loop tracked the ending index of our subarrays. This solution isn't very time efficient because we examine the same elements of the list repeatedly.
+
+<!-- add visualization if time allows -->
+
+Is it possible that an alternative approach can help us achieve simpler or more efficient code? Recall that for this problem we came up with the following subproblems:
+
+1. Finding all subarrays of size k
+2. Determining whether a single subarray contains duplicates
+
+In our current approach we work through a single subarray and check as we move through that subarray whether it contians duplicates. Then we repeat the process with the next subarray. 
+
+Ideally, it would be nice to find all subarrays by only making a single loop through our input array. But if we eliminate the second loop in our current implementation, we no longer update the value that tracks the end index of our subarrays.
+
+What if we flipped our approach so that instead of first confirming we are within a subarray of length k and then checking wehther it contains duplicates, we instead track the indices of every value in our array and use arithmetic to see if any value with more than two instances is less than k?
```

Comment:
```suggestion
What if we flipped our approach so that instead of first confirming we are within a subarray of length k and then checking whether it contains duplicates, we instead track the indices of every value in our array and use arithmetic to see if any value with more than two instances is less than k?
```